exploration/variable-mutability.md

```diff
@@ -135,7 +135,108 @@ _What prior decisions and existing conditions limit the possible design?_
 
 ## Proposed Design
 
-_Describe the proposed solution. Consider syntax, formatting, errors, registry, tooling, interchange._
+Let us introduce a new keyword `input` that allows for the annotation of external variables.
+It works like this:
+
+```
+input {$count :number}
+input {$date :datetime dateStyle=long}
+match {$count}
+when 1 {You received one message on {$date}}
+when * {You received {$count} messages on {$date}}
+```
+
+This effectively replaces the current "hack" of saying `let $foo = {$foo ...}`,
+and provides a way to explicitly declare that a variable is non-local:
+`input {$bar}` doesn't require an annotation.
+
+To correspond with "input",
+let us also change the local variable declaration keyword to be `local` rather than `let`.
+
+At the syntax/data model level,
+`input` or `local` declarations would not be required for all selectors and placeholders,
+but a user-configured validator could of course be stricter.
+
+In the ABNF the change would look like this:
+
+```abnf
+message = [s] *(declaration [s]) body [s]
+declaration = input-declaration / local-declaration
+
+input-declaration = input [s] "{" [s] variable [s annotation] [s] "}"
+input = %x69.6E.70.75.74  ; "input"
+
+local-declaration = local s variable [s] "=" [s] expression
+local = %x6C.6F.63.61.6C  ; "local"
+```
+
+The _expression_ rule can't be used directly in _input-declaration_ because the _variable_ is required.
+
+With this approach, variables are immutable,
+so each may be defined by only one _declaration_.
+The order of declarations does not matter,
+to allow for e.g. `input` annotations to refer to `local` variables.
+
+References to later declarations are not allowed,
+so this is considered an error:
```

Comment:
This confuses me. If the order doesn't matter then forward definitions should work, but then you say that they don't (so the order would matter). I think the first paragraph might be a holdover from when you had input followed by local (the new ABNF allows intermixing: intentional?)

Reply:
We ended up determining that forward-referencing should not be allowed, because if we do allow it, someone not aware of that fact looking at the example message could easily presume that the first `$bar` reference is to an external variable that's shadowed by the `local $bar` declaration.